### PR TITLE
Added s3 module

### PR DESCRIPTION
**Pull request by mayank0202**
## Description

Added s3 module

Screenshot 2024-06-27 at 7 07 32 PM

## Checklist

Please ensure that the following steps are completed before submitting the pull request:

- [x] Code follows the Terraform best practices and style guidelines.
- [x] Changes are appropriately documented, including any necessary updates to README or other documentation files.
- [ ] Unit tests have been added or updated to cover the changes introduced by this pull request.
- [x] Changes have been tested locally and verified to work as expected.
- [x] The code has been reviewed to ensure it aligns with the project's goals and standards.
- [ ] Dependencies and backward compatibility have been considered and addressed if applicable.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## Testing Instructions

ran terraform apply in example folder

### CODE REVIEW

**Review by tsaucier-sf — APPROVED** [review]

File: locals.tf [hunk 1]
```diff
@@ -0,0 +1,19 @@
+locals {
+  bucket_id           = join("", aws_s3_bucket.this[*].id)
```
Comment (vijay-stephen) [inline locals.tf]: `join` is deprecated, please remove

File: locals.tf [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  object_lock_enabled = var.enabled && var.object_lock_configuration != null
```
Comment (vijay-stephen) [inline locals.tf]: try to avoid having `enabled` variable

File: locals.tf [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  versioning_enabled  = var.enabled && var.versioning_enabled
+  partition           = join("", data.aws_partition.current[*].partition)
+  bucket_arn          = "arn:${local.partition}:s3:::${local.bucket_id}"
```
Comment (vijay-stephen) [inline locals.tf]: bucket_arn local variable is required?

File: variables.tf [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    Defaults to "private" for backwards compatibility, but we recommend setting `s3_object_ownership` to "BucketOwnerEnforced" instead.
+    EOT
+}
+
+variable "enabled" {
+  type        = bool
+  default     = true
+  description = ""
+}
+
```
Comment (vijay-stephen) [inline variables.tf]: please remove this variable

**Review by vijay-stephen — CHANGES_REQUESTED** [review]

**Comment by vijay-stephen** [comment]
@mayank0202 Add an example for CRR or SRR

File: locals.tf (hunk 3 above)
Comment (mayank0202) [reply locals.tf]: bucket_arn is used for bucket policy . You can check the data file @vijay-stephen

**Pull request closed without merging**